fix: component type declaration

brandodo opened the pull request:
- Add component typing for Svelte 5

brandodo committed:
fix: component type declaration

theetrain commented:
Thanks, Brandon.

While testing your changes in a Svelte 4 project, I noticed the Cartesian's `Component` prop resolves as type `any` since `Component` is not an export in Svelte 4. 

![image](https://github.com/user-attachments/assets/d7b28441-ae47-4cdb-81af-fa09e4dba23d)

Without these changes, what's the impact? In a Svelte 5 project, I don't see deprecation warnings for `ComponentType` even though it is deprecated. 

![image](https://github.com/user-attachments/assets/284072ac-8794-4027-aea7-df98c8f52c99)

I can also help add feature parity to `CartesianWithRunes` that way you can upgrade to the newer component with newer types. I only need to resolve what's left in #1, and there shouldn't be breaking changes; that way you can upgrade incrementally.

brandodo commented:
@theetrain getting this error when working with a component using Svelte 5 syntax:

```
Error: Type 'Component<Props, {}, "">' is not assignable to type 'ComponentType'.
  Type 'Component<Props, {}, "">' is not assignable to type 'new (options: ComponentConstructorOptions<Record<string, any>>) => SvelteComponent<Record<string, any>, any, any>'.
    Type 'Component<Props, {}, "">' provides no match for the signature 'new (options: ComponentConstructorOptions<Record<string, any>>): SvelteComponent<Record<string, any>, any, any>'. (js)

<Cartesian {props} Component={Switch} />
```

theetrain commented:
Please share a full code example showing how `props` are defined for this "Switch" example. Also does the error occur in your editor, at build time, or someplace else?

brandodo commented:
Occurs during svelte-check run

```
<script>
  /**
   * @typedef Props
   * @property {string} label
   * @property {'off' | 'on'} toggle
   * @property {boolean} asyncSwitch
   * @property {string} formaction
   * @property {import('svelte/elements').HTMLAttributes<HTMLSpanElement>} sliderWrapperSpanAttributes
   * @property {import('svelte/elements').HTMLAttributes<HTMLSpanElement>} sliderSpanAttributes
   * @property {string} [id=crypto.getRandomValues(new Uint32Array(1))[0]]  A unique ID for the input. This defaults to the value of the `name` prop
   * @property {'sm' | 'md'} [size='md'] The size of the input. Using `small` will set the padding of the input to `8px 16px` and `medium` will set the padding of the input to `12px 16px`.
   * @property {import('svelte/elements').HTMLButtonAttributes} buttonAttributes
   * @property {HTMLElement} ref
   */

  /** @type {Props} */
  let {
    label,
    toggle,
    asyncSwitch,
    formaction,
    sliderWrapperSpanAttributes,
    sliderSpanAttributes,
    id,
    size,
    buttonAttributes,
    ref
  } = $props()
</script>
```

theetrain committed:
fix: add conditional component types

- Use `Component` type when Svelte 5 is present, otherwise fallback to `ComponentType`
- Include Svelte 5 in `peerDependencies`
- Patch dependencies

theetrain committed:
chore: patch test dependencies

theetrain committed:
test: update snapshots

theetrain committed:
chore: adjust component type detection

theetrain committed:
fix: adjust types in CartesianWithRunes

- HTML div attributes threw false positives that required overrides.

Details: https://github.com/sveltejs/svelte/issues/13959#issuecomment-2712494762

theetrain commented:
Hi Brandon,

I pushed some additional fixes to your branch, please try it out in your project:

1. Pull latest changes in your `svelte-cartesian` fork
2. Run `npm pack`
3. Install the tarball in your project
4. Observe type checker (you may need to restart the Svelte language server after installing)

Summary of my changes:

1. Add Svelte `^5.0.0` to `peerDependencies` for npm compatibility.
2. Update playwright snapshots
3. Conditionally assign `Component` type to Cartesian's `Component` prop when Svelte 5 is installed; otherwise assign `ComponentType` type to `Component` prop.

References:

- [Svelte 5 conditional type example](https://github.com/sveltejs/language-tools/blob/ddc62b874be822111ce424a9d2ba64f752eaeab4/packages/svelte2tsx/svelte-shims-v4.d.ts#L238).
- [Discord conversation](https://discord.com/channels/457912077277855764/1365353363289673768).

brandodo commented:
@theetrain Tried the steps outlined, however, am still getting type error. Condition for the type seems to evaluate to the truthy result.

theetrain commented:
What are your installed package versions for the following?

```bash
npm ls svelte svelte-check typescript svelte-cartesian
```

Also please share a screenshot of the error.

While I investigate this and follow up, you can temporarily install your commit as a workaround if it unblocks you:

```bash
npm i -D git://github.com/theetrain/svelte-cartesian.git#d9c3edbd21f66d4ed3ac0683ae9f8551e039dd51
```

brandodo commented:
@theetrain 

```
@viralnation/ui-svelte@0.0.0-semantically-released /Users/bong/Desktop/viralnation/ui-svelte
├─┬ @storybook/addon-svelte-csf@5.0.0-next.23
│ ├─┬ svelte-ast-print@0.4.2
│ │ └── svelte@5.20.4 deduped
│ └── svelte@5.20.4 deduped
├─┬ @storybook/svelte-vite@8.5.8
│ ├─┬ svelte-preprocess@5.1.4
│ │ ├── svelte@5.20.4 deduped
│ │ └── typescript@5.7.3 deduped
│ ├── svelte@5.20.4 deduped
│ ├─┬ svelte2tsx@0.7.34
│ │ ├── svelte@5.20.4 deduped
│ │ └── typescript@5.7.3 deduped
│ └── typescript@5.7.3
├─┬ @storybook/svelte@8.5.8
│ └── svelte@5.20.4 deduped
├─┬ @sveltejs/vite-plugin-svelte@4.0.4
│ ├─┬ @sveltejs/vite-plugin-svelte-inspector@3.0.1
│ │ └── svelte@5.20.4 deduped
│ └── svelte@5.20.4 deduped
├─┬ @testing-library/svelte@5.2.7
│ └── svelte@5.20.4 deduped
├─┬ eslint-plugin-svelte@2.46.1
│ ├─┬ svelte-eslint-parser@0.43.0
│ │ └── svelte@5.20.4 deduped
│ └── svelte@5.20.4 deduped
├─┬ prettier-plugin-svelte@3.3.3
│ └── svelte@5.20.4 deduped
├─┬ svelte-cartesian@1.1.5
│ └── svelte@5.20.4 deduped
├─┬ svelte-check@4.1.4
│ ├── svelte@5.20.4 deduped
│ └── typescript@5.7.3 deduped
└── svelte@5.20.4
```
